File: server/auth-contracts/roles.go
package authcontracts

import (
	_ "embed"
	"encoding/json"
)

//go:embed roles.json
var rolesRaw []byte

// RoleWeights is the single role ladder shared by every service.
var RoleWeights = mustLoadRoleWeights()

func mustLoadRoleWeights() map[string]int {
	m := map[string]int{}
	if err := json.Unmarshal(rolesRaw, &m); err != nil {
		panic("auth-contracts: roles.json is invalid: " + err.Error())
	}
	return m
}

func Can(principalRole, requiredRole string) bool {
	have, ok := RoleWeights[principalRole]
	if !ok {
		return false
	}
	need, ok := RoleWeights[requiredRole]
	if !ok {
		return false
	}
	return have >= need
}

File: backend/agent/app/roles.py
"""Role ladder, read from `libs/auth-contracts/roles.json`."""

import json
from pathlib import Path

ROLES_FILE = Path(__file__).resolve().parents[2] / "libs" / "auth-contracts" / "roles.json"
ROLE_WEIGHTS: dict[str, int] = json.loads(ROLES_FILE.read_text())
